### PR TITLE
add koushoku

### LINKED ISSUE #11025: Koushoku
### Source name

Koushoku

### Source link

https://koushoku.org/

### Language

English

### Other details

NSFW (R18+)

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a title with source name.
- [X] I have checked that the extension does not already exist on the [website extensions list](https://tachiyomi.org/extensions/) or the app.
- [X] I have checked that the extension does not already exist by searching the [GitHub repository](https://github.com/tachiyomiorg/tachiyomi-extensions/) and verified it does not appear in the code base.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by rs1703**
Closes #11025 

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [x] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio

### CODE REVIEW

File: src/en/koushoku/src/eu/kanade/tachiyomi/extension/en/koushoku/KoushokuUtils.kt [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package eu.kanade.tachiyomi.extension.en.koushoku
+
+import org.jsoup.nodes.Document
+import org.jsoup.nodes.Element
+
+object KoushokuUtils {
```
Comment (ObserverOfTime) [inline src/en/koushoku/src/eu/kanade/tachiyomi/extension/en/koushoku/KoushokuUtils.kt:6]: If you remove the object, you won't need to import all the methods.

**Review by ObserverOfTime — CHANGES_REQUESTED** [review]

File: src/en/koushoku/src/eu/kanade/tachiyomi/extension/en/koushoku/KoushokuUtils.kt [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    fun getThumbnailURL(document: Document) =
+        document.select(thumbnailSelector).attr("src").trim()
+
+    fun getThumbnailURL(element: Element) =
+        element.select(thumbnailSelector).attr("src").trim()
+
+    fun getArtists(document: Document): String {
+        val artists = document.select(".metadata .artists a")
+        artists.addAll(document.select(".metadata .circles a"))
+        return artists.joinToString(", ") { it.text().trim() }
```
[the remark this answers is not in the document]
Comment (vetleledaal) [reply src/en/koushoku/src/eu/kanade/tachiyomi/extension/en/koushoku/KoushokuUtils.kt:18]: Default seperator ", " for joinToString can be omitted. Pretty sure the jsoup documentation states that .text() is trimmed by default.

Applicable elsewhere as well.

[the remark this answers is not in the document]
Comment (ObserverOfTime) [reply src/en/koushoku/src/eu/kanade/tachiyomi/extension/en/koushoku/KoushokuUtils.kt:18]: Well, if we're gonna be ~~nitpicking~~ doing micro-optimisations, all `.select(...).first()` and `.select(...)` calls that return a single element should be changed to `.selectFirst(...)` instead.

**Comment by rs1703** [comment]
Thanks for the suggestions. I think that's it, I'm not too familiar with kotlin so let me know if it still need fixing.

File: src/en/koushoku/src/eu/kanade/tachiyomi/extension/en/koushoku/Koushoku.kt [hunk 3]
```diff
@@ -149,7 +178,29 @@ class Koushoku : ParsedHttpSource() {
...
+        }
+
+        val pages = document.select(".metadata .pages td:nth-child(2)").text()
+        stringBuilder.append("Pages: $pages\n")
+
+        val size = document.select(".metadata .size td:nth-child(2)").text()
+        stringBuilder.append("Size: $size")
+
+        return stringBuilder.toString()
     }
```
Comment (ObserverOfTime) [inline src/en/koushoku/src/eu/kanade/tachiyomi/extension/en/koushoku/Koushoku.kt:205]:
```suggestion
    private fun getDesc(document: Document) = buildString {
        val magazines = document.select(magazinesSelector)
        if (magazines.isNotEmpty()) {
            append("Magazines: ")
            append(magazines.joinToString { it.text() })
            append("\n")
        }

        val parodies = document.select(".metadata .parodies a")
        if (parodies.isNotEmpty()) {
            append("Parodies: ")
            append(parodies.joinToString { it.text() })
            append("\n")
        }

        val pages = document.selectFirst(".metadata .pages td:nth-child(2)")
        append("Pages: ").append(pages.text()).append("\n")

        val size = document.select(".metadata .size td:nth-child(2)")
        append("Size: ").append(size.text())
    }
```

**Review by ObserverOfTime — CHANGES_REQUESTED** [review]

**Review by ObserverOfTime — APPROVED** [review]

**Pull request merged**